common/utilities/tests/test_timing.py

```diff
@@ -0,0 +1,159 @@
+
+from common.utilities import timing
+from common.utilities.test_utilities import async_test
+
+
+class TestTimeUtilities(TestCase):
+
+    @patch('common.utilities.timing.StopWatch.stop_timer')
+    @patch('logging.info')
+    def test_invoke_with_time(self, log_mock, time_mock):
+        time_mock.return_value = 5
+        self.sleep_method()
+        log_mock.assert_called_with({'method': 'sleep_method', 'duration': 5})
+
+    @patch('common.utilities.timing.StopWatch.stop_timer')
+    @patch('logging.info')
+    def test_exception_thrown_whilst_timing(self, log_mock, time_mock):
+        time_mock.return_value = 5
+        try:
+            self.throw_error_method()
+        except ValueError:
+            pass
+
+        log_mock.assert_called_with({'method': 'throw_error_method', 'duration': 5})
+
+    @patch('time.time')
+    def test_stopwatch(self, time_mock):
+        stopwatch = timing.StopWatch()
+        time_mock.return_value = 0
+        stopwatch.start_timer()
+        time_mock.return_value = 5
+        result = stopwatch.stop_timer()
+
+        self.assertEqual(result, 5)
+
+    @patch('common.utilities.timing.StopWatch.stop_timer')
+    @patch('logging.info')
+    def test_invoke_with_time_rounding(self, log_mock, time_mock):
+        time_mock.return_value = 5.1236
+        self.sleep_method()
+        log_mock.assert_called_with({'method': 'sleep_method', 'duration': 5.124})
+
+    @patch('common.utilities.timing.StopWatch.stop_timer')
+    @patch('logging.info')
+    def test_invoke_with_time_parameters(self, log_mock, time_mock):
+        time_mock.return_value = 5
+        self.take_parameters("whew", 1, [2], {3: 3})
+        log_mock.assert_called_with({'method': 'take_parameters', 'duration': 5})
+
+    @patch('common.utilities.timing.StopWatch.stop_timer')
+    @patch('logging.info')
+    @async_test
+    async def test_invoke_with_time_fake_async_method(self, log_mock, time_mock):
+        time_mock.return_value = 5
+        await self.async_method()
+        log_mock.assert_called_with({'method': 'async_method', 'duration': 5})
+
+    @timing.time_method
+    def sleep_method(self):
+        pass
+
+    @timing.time_method
+    def throw_error_method(self):
+        raise ValueError("Whew")
+
+    @timing.time_method
+    def take_parameters(self, check, one, two, three):
+        assert check is not None
+        assert one is not None
+        assert two is not None
+        assert three is not None
+
+    @timing.time_method_async
+    async def async_method(self):
+        pass
+
+    @patch('logging.info')
+    @async_test
+    async def test_invoke_with_time_async_method(self, log_mock):
+        task = self.actual_sleep()
+
+        time.sleep(2)
+        await task
+        self.assertEqual(len(log_mock.call_args_list), 1)
+
+        call = log_mock.call_args_list[0]
+        args, kwargs = call
+        input_dict = args[0]
+
+        self.assertTrue(math.isclose(input_dict["duration"], 1, rel_tol=0.15))
```

Comment:
I know in previous PRs we've mentioned to avoid the usage of actual `sleep` calls in favor of patching the library, however I found there was varying behavior of the timing decorator between real sleeps and patched sleeps so it felt wrong not to include a full unit test with the realistic/non patched scenario. Any better ideas would be appreciated

Reply:
What was the varying behaviour?

Reply:
It's a bit difficult to describe, but it's what lead me to realise we needed async versions as well, because if I patch the time module and call an `async` method without the `async` decorator it would look like the decorator is working correctly. However without the patch it would log that the async method would take 0s (makes sense since it's a coroutine and will return instantly). So it feels like the non patched test should be included to cover this case

Reply:
Does that not point at the fact that your tests aren't checking that the values returned from the decorated method are as expected? That's an important consideration for the decorator, and doing the check would have highlighted the problem you describe, wouldn't it?

Reply:
Good point, and I will update them to ensure they're returning the expected value as well. Do you think this should negate the need to run with an actual sleep?